File: leetcode/123/MaxProfit.rb
class Integer
    N_BYTES = [42].pack('i').size
    N_BITS = N_BYTES * 16
    MAX = 2 ** (N_BITS - 2) - 1
    MIN = -MAX - 1
end

# @param {Integer[]} prices
# @return {Integer}
def max_profit(prices)
        firstBuy = Integer::MAX;;
        secondBuy = Integer::MAX;;
        firstSell = 0;
        secondSell = 0;
        for i in 0..prices.size-1
            firstBuy = [firstBuy,prices[i]].min;
            firstSell = [firstSell,prices[i] - firstBuy].max;
            secondBuy = [secondBuy,prices[i] - firstSell].min;
            secondSell = [secondSell, prices[i] - secondBuy].max;
            puts(firstBuy,firstSell,secondBuy,secondSell);
            puts("");
        end 
        secondSell;
end

puts max_profit([ 3,3,5,0,0,3,1,4]);

File: leetcode/188/MaxProfit.java
class MaxProfit {
    public static void main(String[] args) {
        Solution s = new Solution();
        System.out.println(s.maxProfit(2, new int[] { 2, 4, 1 }));
        System.out.println(s.maxProfit(2, new int[] { 3, 2, 6, 5, 0, 3 }));
    }
}

class Solution {
    public int maxProfit(int k, int[] prices) {

        if (k == 0 || prices.length < 2)
            return 0;

        if (k > prices.length / 2) {
            int maxProfitII = 0;
            for (int i = 1; i < prices.length; ++i)
                if (prices[i] > prices[i - 1])
                    maxProfitII += prices[i] - prices[i - 1];
            return maxProfitII;
        }
        int[] buy = new int[k];
        int[] sell = new int[k];
        for (int i = 0; i < buy.length; i++)
            buy[i] = Integer.MIN_VALUE;
        for (int i = 0; i < prices.length; i++)
            for (int j = k - 1; j >= 0; j--) {
                sell[j] = Math.max(sell[j], buy[j] + prices[i]);
                if (j == 0)
                    buy[j] = Math.max(buy[j], -prices[i]);
                else
                    buy[j] = Math.max(buy[j], sell[j - 1] - prices[i]);
            }
        return sell[k - 1];
    }
}
